File: codes/0654-Maximum-Binary-Tree/Java/main3.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
public class Solution {
    public TreeNode constructMaximumBinaryTree(int[] nums) {
        if(nums==null || nums.length==0) return null;
        Deque<TreeNode> stack = new ArrayDeque<>(nums.length>>1);
        for(int num : nums) {
            TreeNode cur = new TreeNode(num);
            while(!stack.isEmpty() && stack.peek().val<num){
                cur.left = stack.pop();
            }
            if(!stack.isEmpty()){
                stack.peek().right = cur;
            }
            stack.push(cur);
        }
        return stack.peekLast();
    }
}

File: codes/0654-Maximum-Binary-Tree/cpp/main3.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution
{
public:
    TreeNode *constructMaximumBinaryTree(vector<int> &nums)
    {
        vector<TreeNode *> v;
        for (int num : nums)
        {
            TreeNode *cur = new TreeNode(num);
            while (!v.empty() && v.back()->val < num)
            {
                cur->left = v.back();
                v.pop_back();
            }
            if (!v.empty())
            {
                v.back()->right = cur;
            }
            v.push_back(cur);
        }
        return v.front();
    }
};